ath79: move swconfig support into modules

luizluca opened the pull request:
Ath79 have two phy drives, one that is the swconfig driver and cannot be used with the upstream DSA driver:

- /sys/bus/mdio_bus/drivers/Qualcomm Atheros 8337 internal PHY (a.k.a. CONFIG_AT803X_PHY)
- /sys/bus/mdio_bus/drivers/Atheros AR8216!AR8236!AR8316 (a.k.a. CONFIG_AR8216_PHY)

As a first step to migrate ath79 to DSA (where possible), swconfig support were converted into modules, including its PHY drivers. The Ethernet driver was also converted into a module to postpone the mdio bus initialization until all modules were loaded.

This PR was tested on some TP-Link devices, both in normal and failsafe operation. It would be nice to expand the tests (and the kernel/package cleanup) to other devices that uses different swconfig switches.

luizluca committed:
ath79: drop uneeded packages in some TP-Link devices

Some (previously built-in) modules are not used by all devices. We can
refine the package section on each device.

This was applied only to TP-Link TL-WDR3600v1, TL-WD4300v1, Archer C7v4
and Archer C7v5. However, many more devices might benefit from a similar
cleanup.

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

neheb commented:
I'll note swconfig can't go away from the target. There are realtek switches with swconfig drivers.

john-tho commented on target/linux/ath79/modules.mk:
```diff
@@ -14,3 +14,70 @@ define KernelPackage/leds-reset/description
+$(eval $(call KernelPackage,phy-ar8xxx))
+
+
+define KernelPackage/phy-ip17xx
```
I saw there are already kernel packages for [`kmod-switch-ip17xx`](https://github.com/luizluca/openwrt/blob/bbc7de3b164f43f44feff9ef8e72a2b475953eec/package/kernel/linux/modules/netdevices.mk#L405) (used by one ramips device) and [`kmod-switch-ar8xxx`](https://github.com/luizluca/openwrt/blob/bbc7de3b164f43f44feff9ef8e72a2b475953eec/package/kernel/linux/modules/netdevices.mk#LL517C35-L517C35) (unused). Not sure how well it would work (might be a horrible idea), but, if you only need minor differences to those packages; example, change the `AutoLoad` priority, you might be able to save some duplication (at the expense of clarity) by calling those already-defined packages.
Probably best to wait for some feedback on this idea (calling existing packages) before you spend too much time on it.

Example:
I had to remove my `openwrt/tmp` folder, to get the build system to regenerate `tmp/.config-package.in` 
```
define KernelPackage/switch-ip17xx-18
  $(call KernelPackage/switch-ip17xx)
  CONFLICTS:=kmod-switch-ip17xx
  TITLE+= (autoload priority 18)
  AUTOLOAD:=$(call AutoLoad,18,ip17xx,1)
endef

define KernelPackage/switch-ip17xx-18/description
  $(call KernelPackage/switch-ip17xx/description)
endef

$(eval $(call KernelPackage,switch-ip17xx-18))
```

`CONFIG_AT803X_PHY` is widely used, so you may be able to put that in netdevices

luizluca replied:
> I saw there are already kernel packages for [`kmod-switch-ip17xx`](https://github.com/luizluca/openwrt/blob/bbc7de3b164f43f44feff9ef8e72a2b475953eec/package/kernel/linux/modules/netdevices.mk#L405) (used by one ramips device) and [`kmod-switch-ar8xxx`](https://github.com/luizluca/openwrt/blob/bbc7de3b164f43f44feff9ef8e72a2b475953eec/package/kernel/linux/modules/netdevices.mk#LL517C35-L517C35) (unused). Not sure how well it would work (might be a horrible idea), but, if you only need minor differences to those packages; example, change the `AutoLoad` priority, you might be able to save some duplication (at the expense of clarity) by calling those already-defined packages. Probably best to wait for some feedback on this idea (calling existing packages) before you spend too much time on it.

How I missed those! Thanks! I though I did a grep looking for the symbol but I might have done something wrong. I'll move them to netdevices. I just "synced" the priority close to similar drivers and increased the ethernet driver even further.

> Example: I had to remove my `openwrt/tmp` folder, to get the build system to regenerate `tmp/.config-package.in`

I also noticed that I had to remove openwrt/tmp to get them in the menu. Is that expected? It looks like the build script is not keeping track of target/*/modules.mk
 
> `CONFIG_AT803X_PHY` is widely used, so you may be able to put that in netdevices

I'll migrate to netdevices.

john-tho commented:
> CONFIG_SWCONFIG_LEDS but I have to investigate if kmod-swconfig

Yes, `swconfig.c` does include `swconfig-leds.c`: https://github.com/openwrt/openwrt/blob/6892603efa11603c6f365bee1cf94e98336d72aa/target/linux/generic/files/drivers/net/phy/swconfig.c#L32, which is wrapped in an `ifdef` for `CONFIG_SWCONFIG_LEDS`: https://github.com/openwrt/openwrt/blob/6892603efa11603c6f365bee1cf94e98336d72aa/target/linux/generic/files/drivers/net/phy/swconfig_leds.c#L13

You could extend `kmod-swconfig` `$(call KernelPackage/swconfig)`, or create a new package, something like `swconfig-with-leds`.

---
Not a problem now, but eventually there may be an issue with the buildbot images for devices that use an initramfs, in that `initramfs` images (only needed for first install, or testing) will not respect `DEVICE_PACKAGES`, with [`CONFIG_BUILDBOT`](https://github.com/openwrt/openwrt/blob/4c5a9da8699a7982b8f03b28561f955d9d1313f1/scripts/target-metadata.pl#L274), which includes:
```
CONFIG_TARGET_ALL_PROFILES=y
CONFIG_TARGET_MULTI_PROFILE=y
CONFIG_TARGET_PER_DEVICE_ROOTFS=y
```
As is, this would mean that buildbot initramfs images include both the DSA and swconfig driver modules. Could workaround this with a subtarget.

Commit with some more details: https://github.com/openwrt/openwrt/commit/1984a6bbcaac95c439dd4a6eba2f78c54e9be215.

luizluca commented:
> > CONFIG_SWCONFIG_LEDS but I have to investigate if kmod-swconfig
> 
> Yes, `swconfig.c` does include `swconfig-leds.c`:

It does not select CONFIG_SWCONFIG when set so I just put it in ath79 config.


> Not a problem now, but eventually there may be an issue with the buildbot images for devices that use an initramfs, in that `initramfs` images (only needed for first install, or testing) will not respect `DEVICE_PACKAGES`, with [`CONFIG_BUILDBOT`](https://github.com/openwrt/openwrt/blob/4c5a9da8699a7982b8f03b28561f955d9d1313f1/scripts/target-metadata.pl#L274), which includes:
> 
> ```
> CONFIG_TARGET_ALL_PROFILES=y
> CONFIG_TARGET_MULTI_PROFILE=y
> CONFIG_TARGET_PER_DEVICE_ROOTFS=y
> ```
> 
> As is, this would mean that buildbot initramfs images include both the DSA and swconfig driver modules. Could workaround this with a subtarget.
> 
> Commit with some more details: [1984a6b](https://github.com/openwrt/openwrt/commit/1984a6bbcaac95c439dd4a6eba2f78c54e9be215).

It will definitely affect this PR as it is. If I understand, it uses the plain shared rootfs to build the initramfs image, not installing any device-specific packages. If I install kmod-phy-at803x and kmod-switch-ar8xxx as default, the first one loaded (kmod-switch-ar8xxx as in this PR) will have precedence and it will break DSA. It would need some kind of code adaptation in kmod-switch-ar8xxx to detect a dsa device-tree entries and defer probing or the opposite, at803x detecting swconfig and defering. Alternatively, we could do some userland driver rebind, but it would need to deal with both normal and failsafe.

This limitation would break anything that is required to boot a specific device. Normal initfs support some type of appended image. Can't we use something like this for initramfs images? Just extract all the packages and append as a device-extra rootfs. Sadly, in this PR specific case, having both drivers is still not enough. It would need to have a append that removes files from removed packages. Let's wait a little bit and see if someone comes with a better solution.

luizluca commented:
I just pushed an update:

- using kmod-switch-{ar8xxx,ip17xx} from netdevices.mk instead of kmod-phy-{ar8xxx,ip17xx} from target modules.mk
- migrated kmod-phy-at803x to netdevices.mk
- added CONFIG_AR8216_PHY_LEDS and CONFIG_SWCONFIG_LEDS to target config as they do not break compiling them as module.
- rearranged module priorities to use existing kmd-switch-* packages.

john-tho commented:
> It will definitely affect this PR as it is. If I understand, it uses the plain shared rootfs to build the initramfs image, not installing any device-specific packages.

Yes, we do not have to deal with this now, but we need to consider it for the future. We can continue as you have started, with swconfig drivers prioritised.
I think (once we have to deal with buildbots):
- Minimum effort; move (or copy+rename) DSA-converted devices (which can use initramfs) to another subtarget (where the swconfig driver packages are not included in subtarget level `DEFAULT_PACKAGES`)
- Ideal world; swconfig/DSA are interchangeable, load whichever you prefer. Technical challenges; bindings working for both ([DT can be modified live, or overlays used](https://elixir.bootlin.com/linux/latest/source/drivers/of/unittest.c)), selectable phy drivers, and userspace OpenWrt networking working with both swconfig `switch-vlans` and DSA device-per-port.

---
For the next time you need to force-push, rebase for your `mdio-bus-mux` typo-fix that got picked up: https://github.com/openwrt/openwrt/commit/1e4bc13eaa3fdba897ca1721b2bfe9f2dbb30770

Cheers

abajk commented on target/linux/ath79/Makefile:
```diff
@@ -14,6 +14,8 @@ include $(INCLUDE_DIR)/target.mk
 	kmod-gpio-button-hotplug swconfig \
-	kmod-ath9k uboot-envtools
+	kmod-ath9k uboot-envtools \
+	kmod-switch-ar8xxx kmod-switch-ip17xx \
```
Kmod-switch-ip17xx is only needed on the generic subtarget.

luizluca replied:
done

luizluca commented:
> Not a problem now, but eventually there may be an issue with the buildbot images for devices that use an initramfs, in that `initramfs` images (only needed for first install, or testing) will not respect `DEVICE_PACKAGES`, with [`CONFIG_BUILDBOT`](https://github.com/openwrt/openwrt/blob/4c5a9da8699a7982b8f03b28561f955d9d1313f1/scripts/target-metadata.pl#L274), which includes:
> 
> ```
> CONFIG_TARGET_ALL_PROFILES=y
> CONFIG_TARGET_MULTI_PROFILE=y
> CONFIG_TARGET_PER_DEVICE_ROOTFS=y
> ```
> 
> As is, this would mean that buildbot initramfs images include both the DSA and swconfig driver modules. Could workaround this with a subtarget.
> 
> Commit with some more details: [1984a6b](https://github.com/openwrt/openwrt/commit/1984a6bbcaac95c439dd4a6eba2f78c54e9be215)

I wonder why initramfs does not respect TARGET_PER_DEVICE_ROOTFS. Is it some design limitation or just nobody cared to adapt it?

Wouldn't it be something like:

```
initramfs-build
   if multi_rootfs
      use target-dir-XXXXX.cpio (creating if it does not exist)
      call kernel to update vmlinux-initramfs.bin.XXXXX
   else
      go as usual
   endif
```

luizluca commented:
I think it is time to review for merging.

robimarko commented:
@luizluca Initramfs images dont respect `TARGET_PER_DEVICE_ROOTFS` as its a design limitation, as we are building only single initramfs image per subtarget.

luizluca commented:
> @luizluca Initramfs images dont respect `TARGET_PER_DEVICE_ROOTFS` as its a design limitation, as we are building only single initramfs image per subtarget.

Would the cost of building a different rootfs for each group of packages be acceptable? I can try to write a patch.

From what I checked, kernel is build beforehand, both with and without initramfs. Maybe we can move the initramfs out of that part and call it for each target-dir-XXXX as we do for other images. I believe we need to create a new target-dir-XXXX.initramfs as it requires a couple of extra files (or only the /init). How about that approach?

robimarko commented:
Kernel with initramfs is built right before packaging the sysupgrade images, so we would have to rebuild initramfs for each device, this would probably take a lot of time for targets with a lot of devices.

luizluca commented:
> Kernel with initramfs is built right before packaging the sysupgrade images, so we would have to rebuild initramfs for each device, this would probably take a lot of time for targets with a lot of devices.
> 
> For me it would be worth to use up some more build time but then have the ability to not have to include eMMC specific packages on all boards, or on QCA targets be able to include the proper BDF even in initramfs.

We can share the same initramfs that uses the same package set. This is how rootfs per device works.

robimarko commented:
If you are interested in adding support for it, then please go ahead as it affects targets I work on as well and has been one of the current annoying limitations of the build system.

luizluca commented:
> If you are interested in adding support for it, then please go ahead as it affects targets I work on as well and has been one of the current annoying limitations of the build system.

It is still a draft but it already works for my case: https://github.com/openwrt/openwrt/pull/12959

luizluca commented:
Rebased to remove commit already merged into main.

Independently from https://github.com/openwrt/openwrt/pull/12959 (I need some help there), I see no harm to merge this as all new modules are still installed by default. It does allow some image cleanup for each device (as I did for C7v4).

mpratt14 commented:
I'd like to note that 803x are chip models and therefore NOT internal, although the driver may also be handling similarly designed internal switches

do we know for sure that it does handle internal switches as well? I was about to suggest that we can have that new module only enabled for specific devices

Either way, the description confused me

https://media.digikey.com/pdf/data%20sheets/csr%20pdfs/ar8035_ds_(atheros)_mar2011.pdf

abajk commented:
@luizluca Some parts of this PR were merged in 7e5f971126. Can you rebase this PR?

luizluca commented:
Sure. I'm away from my workstation but I'll take a look ASAP.

luizluca committed:
ath79: convert swconfig drivers into modules

In a first step to gradually adopt DSA switch, all swconfig modules must
be optional, loaded as modules when needed.

AT803X_PHY was also converted to drop it from swconfig devices as
swconfig uses AR8216_PHY/ar8xxx. Once we convert a device to DSA, we
can do the opposite and drop ar8xxx and include at803x.

AG71XX (ethernet driver) was also converted into a module to postpone
mdio bus initialization. If not, mdio.0 would be registered with the
first compatible driver, normally "Generic PHY".

All modules are still installed by default maintainig their previous
availability. However, they should be removed from default selections
and included only on the group of device they are needed.

There are still other PHY drivers builtin that could follow the same
path as AT803X_PHY and AR8216_PHY. The /sys/bus/mdio_bus/drivers/
directory might give a hint of what to convert into modules (and
selectively require on each needed devices).

Tested on TP-Link TL-WDR3600v1, TL-WD4300v1 and Archer C7v5, both in
normal and failsafe mode.

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

luizluca commented:
> Sure. I'm away from my workstation but I'll take a look ASAP.

Just a minor update: I'm working on it. I just need to test the rebased branch on multiple devices/situations.

luizluca committed:
kernel: do not sort autoload modules

The order which modules are loaded might matter and forcing the
alphabetic order might not match that sequence (like in
ar7xx_legacy_mdio, ar7xx_legacy).

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

luizluca committed:
kernel: netdevices: reorder ag71xx_legacy autoload

ag71xx_legacy_mdio is needed to properly probe the device. If
ag71xx_legacy is loaded before, it will defer with:

ag71xx-legacy 19000000.eth: Could not connect to PHY device. Deferring
probe.

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

luizluca commented:
> @luizluca Some parts of this PR were merged in [7e5f971](https://github.com/openwrt/openwrt/commit/7e5f971126ebe59f87414a7e08d02344c4acb7ab). Can you rebase this PR?

Rebased done. I also expanded the device cleanup to other TP-Link devices I have in hand.

There are still a couple of builtin drivers that should be converted into modules and dozens of devices that need a cleanup.

PolynomialDivision commented:
Just stumbled on this, looks good as a prep step for DSA. Is the conversion of the remaining built-in drivers still ongoing, or do you consider this PR finished?
